elki/src/main/java/elki/math/statistics/dependence/MCDEDependenceMeasure.java

```diff
@@ -0,0 +1,249 @@
+   */
+  protected double alpha = 0.5;
+
+  /**
+   * Parameter that specifies the size of the marginal restriction. Note that in the original paper
+   * alpha = beta and as such there is no explicit distinction between the parameters.
+   */
+  protected double beta = 0.5;
+
+  /**
+   * Random generator.
+   */
+  protected RandomFactory rnd;
+
+  /**
+   * Statistical Test returning p-value tailored to MCDE Framework.
+   */
+  protected MCDETest mcdeTest;
+
+  /**
+   * Constructor.
+   */
+  public MCDEDependenceMeasure(int m, double alpha, double beta, RandomFactory rnd, MCDETest mcdeTest) {
+    this.m = m;
+    this.alpha = alpha;
+    this.beta = beta;
+    this.rnd = rnd;
+    this.mcdeTest = mcdeTest;
+  }
+
+  /**
+   * Data Slicing
+   *
+   * @param len         No of data instances
+   * @param nonRefIndex Index (see correctedRank()) computed for the dimension that is not the reference dimension
+   * @return Array of booleans that states which instances are part of the slice
+   */
+  protected boolean[] randomSlice(int len, MCDETest.RankStruct[] nonRefIndex) {
+    final Random random = rnd.getSingleThreadedRandom();
+    boolean slice[] = new boolean[len];
+    Arrays.fill(slice, true);
+
+    // According to the actual formula it should be Math.ceil(Math.pow(this.alpha, 1.0) * len).
+    // The exponent in the multivariate case should be the no. of dimensions - 1 which in the
+    // bivariate case is always simply 1.
+    final int slizeSize = (int) Math.ceil(this.alpha * len);
+    final int start = random.nextInt(len - slizeSize);
+    final int end = start + slizeSize;
+
+    for(int j = 0; j < start; j++) {
+      slice[nonRefIndex[j].index] = false;
+    }
+
+    for(int j = end; j < len; j++) {
+      slice[ nonRefIndex[j].index] = false;
+    }
+
+    return slice;
+  }
+
+  /**
+   * Implements dependence from DependenceMeasure superclass. Corresponds to Algorithm 4 in source paper.
+   *
+   * @param adapter1 First data adapter
+   * @param data1    First data set
+   * @param adapter2 Second data adapter
+   * @param data2    Second data set
+   * @param <A>      Numeric data type, such as double
+   * @param <B>      Numeric data type, such as double
+   * @return MCDE result
+   */
+  @Override public <A, B> double dependence(final NumberArrayAdapter<?, A> adapter1, final A data1, final NumberArrayAdapter<?, B> adapter2, final B data2) {
+    final Random random = rnd.getSingleThreadedRandom();
+    final int len = adapter1.size(data1);
+
+    if(len != adapter2.size(data2)) {
+      throw new AbortException("Size of both arrays must match!");
+    }
+
+    final MCDETest.RankStruct[] index_0 = mcdeTest.corrected_ranks(adapter1, data1, len);
+    final MCDETest.RankStruct[] index_1 = mcdeTest.corrected_ranks(adapter2, data2, len);
+
+    double mwp = 0;
+
+    for(int i = 0; i < this.m; i++) {
+      int r = random.nextInt(2);
+      MCDETest.RankStruct[] ref_index;
```

Comment:
`ref_index = flip ? index_1 : index_0;`
`other_index = flip ? index_0 : index_1;`
Some people have issues with ternary statements, but they often are much more concise and usually produce shorter bytecode and even faster code in the end. Alternatively, the case distinction could be moved below `start` and `end` so these two variables are not used at all.